sos/plugins/juju_controller.py

```diff
@@ -0,0 +1,157 @@
+    # Use files instead of packages here because there is no package installed
+    #  on a juju controller that uniquely identifies it:
+    files = "/var/lib/juju/db"
+
+    option_list = [
+        (
+            "export-mongodb",
+            "Export mongodb collections as json files",
+            "",
+            False,
+        )
+    ]
+
+    def export_mongodb(self, juju_machine_id, dbpass):
+        # We are gonna grab the output of the mongo command
+        # and do some regexp and replace()/split() it to a list.
+        # Usually when connecting with ssl on localhost, the cli send some
+        # output which unfortunately ends up in stdin instead of stderr
+        # and we need to do this extra work to parse it.
+        collections_query = ("mongo 127.0.0.1:37017/juju "
+                             + "--authenticationDatabase admin "
+                             + "--ssl --sslAllowInvalidCertificates --quiet "
+                             + "--username %s --password %s "
+                             % (juju_machine_id, dbpass)
+                             + "--eval 'db.getCollectionNames();'")
+        collections_output = os.popen(collections_query).read()
+        collections_regexp = re.compile(r"(?ms)^\[(.*)\]")
+        collections = collections_regexp.findall(collections_output)[0]
+        subs = [
+            ('\"', ''),
+            (',', ''),
+        ]
+        for search, replace in subs:
+            collections = collections.replace(search, replace)
+        collections = collections.split()
+
+        # We don't collect txns.log, statuseshistory, metrics, txns due to size
+        collections_ignore = ['txns.log', 'statuseshistory', 'metrics', 'txns']
+        for item in collections_ignore:
+            if item in collections:
+                collections.remove(item)
+
+        for collection in collections:
+            filename = "mongoexport_collection_{}.json".format(collection)
+            # This will execute a mongoexport command, note it will add the
+            # password to the sos log, however we remove it from the filename
+            mongoexport_cmd = (
+                "mongoexport --host 127.0.0.1 --port 37017 --db juju "
+                + "--authenticationDatabase admin --ssl "
+                + "--sslAllowInvalidCertificates --username '{}' --password "
+                + "'{}' --collection {} --jsonArray"
+            )
+            self.add_cmd_output(
+                mongoexport_cmd.format(juju_machine_id, dbpass, collection),
+                suggest_filename=filename,
+            )
+
+    def get_juju_machine_id(self):
+        # get the juju machine name machine-X
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                match = re.match(r"(machine-\d+)", dirname)
+                if match:
+                    return match.group(1)
+        return None
+
+    def get_jujudb_pass(self, juju_machine_id):
```

Comment:
We should have the user provide a password to us, rather than trying to dig it up ourselves. Like the `postgresql` plugin, in order for us to get a db dump the user *must* supply the password via `-k postgresql.password=foobar`

Reply:
The reason we fetch the password is due to the fact that the password is always automatically generated on installation and the user never sets anything here and doesn't need to know it as a general rule.
We would prefer to have the password being fetched automatically due to the fact above.
Let me know if this is enough reasoning to leave it automated, if not we can always create the parameter and instruct the users how to fetch/set the password.

Reply:
IMHO this is still safe (up to two concerns). The password is used just internally and not leaked anywhere else than:

- in process listing during the `mongo` command execution; so any unprivileged user can get the password (is there a way like env.variable to hide it? I am afraid not..)
- in output file generated by sosreport, the password will be stored and provided to any support representative - until `suggest_filename` is used (`get_option` has it, `export_mongodb` will be updated and then it should use `suggest_filename` as well)

Reply:
Hi,

For your 1st concern, there is a way to pass the password parameter empty and wait for the mongodb prompt to ask for it, I can see if there is a way to improve that part but it will be a little bit more complex.

For your second concern, I'm sorry but I'm not understanding what do you mean, we are using suggest_filename on both add_cmd_output.